# <font color=blue>Assignments for "Understanding The Relationship"</font>

To close out this lesson, you're going to do three assignments. For the first assignment, you'll write up a short answer to a question in a Gist file.  For the second two assignments, you'll do your work in Jupyter notebooks, and you should link to those notebooks in the same Gist file.

Please submit a single Gist file containing the answer to first assignment, plus links for second two.

## 1. Interpretion and signficance

Suppose that we would like to know how much families in the US are spending on recreation annually. We estimated the following model:

$$ expenditure = 873 + 0.0012annual\_income + 0.00002annual\_income^2 - 223.57have\_kids $$

*expenditure* is the annual spending on recreation in US dollars, *annual_income* is the annual income in US dollars and *have_kids* is a dummy variable indicating the families with children. Interpret the estimated coefficients. What additional statistics should be given in order to make sure that your interpretations make sense statistically. Write up your answer and save in a Gist. 

## 2. Weather model

In this exercise, you'll work with the historical temperature data from the previous lesson. To complete this assignment, submit a link in the gist file to the Jupyter notebook containing your solutions to the following tasks:

In [ ]:
import numpy as np
import matplotlib.pyplot as plt
import seaborn as sns
import statsmodels.api as sm

/usr/local/lib/python3.7/dist-packages/statsmodels/tools/_testing.py:19: FutureWarning: pandas.util.testing is deprecated. Use the functions in the public API at pandas.testing instead.
  import pandas.util.testing as tm


In [ ]:
import pandas as pd
weather_df = pd.read_csv("https://djl-lms-assets.s3.eu-central-1.amazonaws.com/datasets/weatherHistory.csv")
weather_df.head()

Formatted Date  ...                      Daily Summary
0  2006-04-01 00:00:00.000 +0200  ...  Partly cloudy throughout the day.
1  2006-04-01 01:00:00.000 +0200  ...  Partly cloudy throughout the day.
2  2006-04-01 02:00:00.000 +0200  ...  Partly cloudy throughout the day.
3  2006-04-01 03:00:00.000 +0200  ...  Partly cloudy throughout the day.
4  2006-04-01 04:00:00.000 +0200  ...  Partly cloudy throughout the day.

[5 rows x 12 columns]

- Build a linear regression model where your target variable is the difference between the *apparenttemperature* and the *temperature*. As explanatory variables, use *humidity* and *windspeed*. Now, estimate your model using OLS. Are the estimated coefficients statistically significant? Are the signs of the estimated coefficients in line with your previous expectations? Interpret the estimated coefficients. What are the relations between the target and the explanatory variables? 

In [ ]:
X = weather_df[['Humidity', 'Wind Speed (km/h)']]

y = weather_df['Temperature (C)'] - weather_df['Apparent Temperature (C)']


In [ ]:
X = sm.add_constant(X)

results = sm.OLS(y, X).fit()

results.summary()

<class 'statsmodels.iolib.summary.Summary'>
"""
                            OLS Regression Results                            
==============================================================================
Dep. Variable:                      y   R-squared:                       0.288
Model:                            OLS   Adj. R-squared:                  0.288
Method:                 Least Squares   F-statistic:                 1.949e+04
Date:                Sun, 30 May 2021   Prob (F-statistic):               0.00
Time:                        21:50:17   Log-Likelihood:            -1.7046e+05
No. Observations:               96453   AIC:                         3.409e+05
Df Residuals:                   96450   BIC:                         3.409e+05
Df Model:                           2                                         
Covariance Type:            nonrobust                                         
=====================================================================================
                        coef    std err          t      P>|t|      [0.025      0.975]
-------------------------------------------------------------------------------------
const                -2.4381      0.021   -115.948      0.000      -2.479      -2.397
Humidity              3.0292      0.024    126.479      0.000       2.982       3.076
Wind Speed (km/h)     0.1193      0.001    176.164      0.000       0.118       0.121
==============================================================================
Omnibus:                     3935.747   Durbin-Watson:                   0.264
Prob(Omnibus):                  0.000   Jarque-Bera (JB):             4613.311
Skew:                           0.478   Prob(JB):                         0.00
Kurtosis:                       3.484   Cond. No.                         88.1
==============================================================================

Warnings:
[1] Standard Errors assume that the covariance matrix of the errors is correctly specified.
"""

P values ​​for all coefficients are 0, which means they are all statistically significant. Both coefficients have a positive sign, while the humidity coefficient is greater than the wind speed.

- Next, include the interaction of *humidity* and *windspeed* to the model above and estimate the model using OLS. Are the coefficients statistically significant? Did the signs of the estimated coefficients for *humidity* and *windspeed* change? Interpret the estimated coefficients.

In [ ]:
X['humidity*windspeed'] = X['Humidity'] * X['Wind Speed (km/h)']

results = sm.OLS(y, X).fit()

results.summary()

<class 'statsmodels.iolib.summary.Summary'>
"""
                            OLS Regression Results                            
==============================================================================
Dep. Variable:                      y   R-squared:                       0.341
Model:                            OLS   Adj. R-squared:                  0.341
Method:                 Least Squares   F-statistic:                 1.666e+04
Date:                Sun, 30 May 2021   Prob (F-statistic):               0.00
Time:                        21:50:19   Log-Likelihood:            -1.6669e+05
No. Observations:               96453   AIC:                         3.334e+05
Df Residuals:                   96449   BIC:                         3.334e+05
Df Model:                           3                                         
Covariance Type:            nonrobust                                         
======================================================================================
                         coef    std err          t      P>|t|      [0.025      0.975]
--------------------------------------------------------------------------------------
const                 -0.0839      0.033     -2.511      0.012      -0.149      -0.018
Humidity              -0.1775      0.043     -4.133      0.000      -0.262      -0.093
Wind Speed (km/h)     -0.0905      0.002    -36.797      0.000      -0.095      -0.086
humidity*windspeed     0.2971      0.003     88.470      0.000       0.291       0.304
==============================================================================
Omnibus:                     4849.937   Durbin-Watson:                   0.262
Prob(Omnibus):                  0.000   Jarque-Bera (JB):             9295.404
Skew:                           0.378   Prob(JB):                         0.00
Kurtosis:                       4.320   Cond. No.                         193.
==============================================================================

Warnings:
[1] Standard Errors assume that the covariance matrix of the errors is correctly specified.
"""


- P-values ​​are still 0 when we examine them, which means they are statistically significant. In addition, the signs of both Humidity and Wind Speed ​​have now changed to negative and have smaller absolute values ​​than the values ​​we reviewed earlier. Also, the interaction variable has a positive coefficient.

##  3. House prices model

In this exercise, you'll interpret your house prices model. To complete this assignment, submit a link in the gist file to the Jupyter notebook containing your solutions to the following tasks:

In [ ]:
house_prices_df = pd.read_csv("https://djl-lms-assets.s3.eu-central-1.amazonaws.com/datasets/house_prices.csv", sep = ";")
house_prices_df.head()

Id  MSSubClass MSZoning  ...  SaleType  SaleCondition SalePrice
0   1          60       RL  ...        WD         Normal    208500
1   2          20       RL  ...        WD         Normal    181500
2   3          60       RL  ...        WD         Normal    223500
3   4          70       RL  ...        WD        Abnorml    140000
4   5          60       RL  ...        WD         Normal    250000

[5 rows x 81 columns]

- Run your house prices model again and interpret the results. Which features are statistically significant and which are not?

In [ ]:
df_null = house_prices_df.isnull().sum() / len(house_prices_df)
house_prices_df.drop(df_null[df_null > .1].index, axis=1, inplace=True)
house_prices_df.dropna(inplace=True)

In [ ]:
X = house_prices_df.select_dtypes(exclude='object').drop('SalePrice', axis=1)
y = house_prices_df['SalePrice']

In [ ]:
X = sm.add_constant(X)

results = sm.OLS(y, X).fit()

results.summary()

<class 'statsmodels.iolib.summary.Summary'>
"""
                            OLS Regression Results                            
==============================================================================
Dep. Variable:              SalePrice   R-squared:                       0.808
Model:                            OLS   Adj. R-squared:                  0.803
Method:                 Least Squares   F-statistic:                     161.5
Date:                Sun, 30 May 2021   Prob (F-statistic):               0.00
Time:                        22:07:02   Log-Likelihood:                -15881.
No. Observations:                1338   AIC:                         3.183e+04
Df Residuals:                    1303   BIC:                         3.201e+04
Df Model:                          34                                         
Covariance Type:            nonrobust                                         
=================================================================================
                    coef    std err          t      P>|t|      [0.025      0.975]
---------------------------------------------------------------------------------
const          4.344e+05   1.49e+06      0.292      0.770   -2.48e+06    3.35e+06
Id               -0.4159      2.304     -0.180      0.857      -4.937       4.105
MSSubClass     -164.9821     28.463     -5.796      0.000    -220.820    -109.144
LotArea           0.3950      0.102      3.860      0.000       0.194       0.596
OverallQual    1.854e+04   1277.521     14.509      0.000     1.6e+04     2.1e+04
OverallCond    5539.4909   1144.985      4.838      0.000    3293.275    7785.707
YearBuilt       343.8423     78.784      4.364      0.000     189.286     498.399
YearRemodAdd    149.4128     75.737      1.973      0.049       0.833     297.992
MasVnrArea       27.9086      6.117      4.562      0.000      15.908      39.910
BsmtFinSF1       10.6647      2.743      3.888      0.000       5.284      16.046
BsmtFinSF2        1.9829      4.753      0.417      0.677      -7.342      11.308
BsmtUnfSF         0.6779      2.754      0.246      0.806      -4.725       6.081
TotalBsmtSF      13.3256      4.722      2.822      0.005       4.062      22.590
1stFlrSF         11.0700      7.818      1.416      0.157      -4.268      26.408
2ndFlrSF         18.5493      6.803      2.727      0.006       5.203      31.896
LowQualFinSF      0.9313     18.561      0.050      0.960     -35.482      37.345
GrLivArea        30.5506      6.929      4.409      0.000      16.957      44.144
BsmtFullBath   8486.5250   2773.439      3.060      0.002    3045.630    1.39e+04
BsmtHalfBath   2007.5870   4271.709      0.470      0.638   -6372.593    1.04e+04
FullBath       1106.9972   3071.372      0.360      0.719   -4918.378    7132.373
HalfBath      -3940.6439   2848.081     -1.384      0.167   -9527.970    1646.682
BedroomAbvGr  -9641.7992   1875.348     -5.141      0.000   -1.33e+04   -5962.766
KitchenAbvGr  -2.589e+04   6633.844     -3.903      0.000   -3.89e+04   -1.29e+04
TotRmsAbvGrd   5615.9799   1302.811      4.311      0.000    3060.144    8171.816
Fireplaces     3777.7691   1852.760      2.039      0.042     143.051    7412.487
GarageYrBlt     -48.8651     81.702     -0.598      0.550    -209.147     111.417
GarageCars     1.524e+04   3079.095      4.948      0.000    9195.805    2.13e+04
GarageArea        4.8423     10.400      0.466      0.642     -15.560      25.245
WoodDeckSF       23.7624      8.253      2.879      0.004       7.572      39.952
OpenPorchSF      -6.8007     16.325     -0.417      0.677     -38.827      25.226
EnclosedPorch    11.1010     17.862      0.621      0.534     -23.941      46.143
3SsnPorch        23.2182     32.087      0.724      0.469     -39.730      86.166
ScreenPorch      54.7314     17.391      3.147      0.002      20.614      88.849
PoolArea        -36.8969     23.959     -1.540      0.124     -83.899      10.105
MiscVal          -0.6979      1.909     -0.366   

- Now, exclude the insignificant features from your model. Did anything change?

In [ ]:
df_train_corr = house_prices_df.corr()['SalePrice'][:-1] # -1 because the latest row is SalePrice
golden_features_list = df_train_corr[abs(df_train_corr) >0.4].sort_values(ascending=False)
print("There is {} strongly correlated values with SalePrice:\n{}".format(len(golden_features_list), golden_features_list))

There is 13 strongly correlated values with SalePrice:
OverallQual     0.783546
GrLivArea       0.711706
GarageCars      0.640154
GarageArea      0.607535
1stFlrSF        0.604714
TotalBsmtSF     0.602042
FullBath        0.569313
TotRmsAbvGrd    0.551821
YearBuilt       0.504297
YearRemodAdd    0.501435
GarageYrBlt     0.481730
MasVnrArea      0.465811
Fireplaces      0.445434
Name: SalePrice, dtype: float64


In [ ]:
X_sig= X[['OverallQual', 'GrLivArea', 'GarageCars', 'GarageArea', 'TotalBsmtSF', 'FullBath', 'YearBuilt','YearRemodAdd',
          'TotRmsAbvGrd','1stFlrSF', 'BsmtFullBath', 'BedroomAbvGr','KitchenAbvGr','GarageYrBlt','Fireplaces']]

In [ ]:
results = sm.OLS(y, X_sig).fit()

results.summary()

<class 'statsmodels.iolib.summary.Summary'>
"""
                                 OLS Regression Results                                
=======================================================================================
Dep. Variable:              SalePrice   R-squared (uncentered):                   0.967
Model:                            OLS   Adj. R-squared (uncentered):              0.967
Method:                 Least Squares   F-statistic:                              2591.
Date:                Sun, 30 May 2021   Prob (F-statistic):                        0.00
Time:                        23:22:50   Log-Likelihood:                         -15965.
No. Observations:                1338   AIC:                                  3.196e+04
Df Residuals:                    1323   BIC:                                  3.204e+04
Df Model:                          15                                                  
Covariance Type:            nonrobust                                                  
================================================================================
                   coef    std err          t      P>|t|      [0.025      0.975]
--------------------------------------------------------------------------------
OverallQual   1.999e+04   1247.551     16.020      0.000    1.75e+04    2.24e+04
GrLivArea       40.9789      4.420      9.271      0.000      32.308      49.650
GarageCars    1.684e+04   3149.169      5.348      0.000    1.07e+04     2.3e+04
GarageArea      24.6733     10.760      2.293      0.022       3.565      45.781
TotalBsmtSF     26.4693      6.118      4.326      0.000      14.466      38.472
FullBath      5328.1873   2715.032      1.962      0.050       1.951    1.07e+04
YearBuilt      188.1998     69.570      2.705      0.007      51.720     324.679
YearRemodAdd    83.6212     57.946      1.443      0.149     -30.055     197.298
TotRmsAbvGrd  6133.1230   1349.351      4.545      0.000    3486.022    8780.224
1stFlrSF        -1.1048      6.506     -0.170      0.865     -13.867      11.658
BsmtFullBath   1.43e+04   2148.251      6.655      0.000    1.01e+04    1.85e+04
BedroomAbvGr -8669.9422   1891.605     -4.583      0.000   -1.24e+04   -4959.070
KitchenAbvGr -5.305e+04   6309.030     -8.408      0.000   -6.54e+04   -4.07e+04
GarageYrBlt   -297.1098     79.547     -3.735      0.000    -453.162    -141.058
Fireplaces    4629.1948   1878.781      2.464      0.014     943.479    8314.910
==============================================================================
Omnibus:                      563.988   Durbin-Watson:                   1.968
Prob(Omnibus):                  0.000   Jarque-Bera (JB):            75167.341
Skew:                          -0.903   Prob(JB):                         0.00
Kurtosis:                      39.675   Cond. No.                     2.59e+04
==============================================================================

Warnings:
[1] Standard Errors assume that the covariance matrix of the errors is correctly specified.
[2] The condition number is large, 2.59e+04. This might indicate that there are
strong multicollinearity or other numerical problems.
"""

R-square has improved significantly and is approaching the desired result. (It is desired that the value of R-squared be very close to 1)

- Interpret the statistically significant coefficients by quantifying their relations with the house prices. Which features have more prominent effect on the house prices?

Some coefficients did not change much, while others changed a lot. For example, the coefficient of the variable "OverallQual", which had the highest effect on the target variable, increased from 18540 to 19990. "YearBuilt" decreased from 343.8423 to 188.1998.

- Do the results sound reasonable to you? If not, try to explain the potential reasons.

Yes, I think they are reasonable.